axelrod/tests/unit/test_tournament.py

```diff
@@ -394,30 +397,128 @@ def test_play_matches(self, turns):
             players=self.players,
             game=self.game,
             repetitions=self.test_repetitions)
-        matches = {
-            (0, 0): axelrod.Match((axelrod.Cooperator(), axelrod.Cooperator()), turns=turns),
-            (0, 1): axelrod.Match((axelrod.Cooperator(), axelrod.TitForTat()), turns=turns),
-            (0, 2): axelrod.Match((axelrod.Cooperator(), axelrod.Defector()), turns=turns),
-            (0, 3): axelrod.Match((axelrod.Cooperator(), axelrod.Grudger()), turns=turns),
-            (0, 4): axelrod.Match((axelrod.Cooperator(), axelrod.GoByMajority()), turns=turns),
-            (1, 1): axelrod.Match((axelrod.TitForTat(), axelrod.TitForTat()), turns=turns),
-            (1, 2): axelrod.Match((axelrod.TitForTat(), axelrod.Defector()), turns=turns),
-            (1, 3): axelrod.Match((axelrod.TitForTat(), axelrod.Grudger()), turns=turns),
-            (1, 4): axelrod.Match((axelrod.TitForTat(), axelrod.GoByMajority()), turns=turns),
-            (2, 2): axelrod.Match((axelrod.Defector(), axelrod.Defector()), turns=turns),
-            (2, 3): axelrod.Match((axelrod.Defector(), axelrod.Grudger()), turns=turns),
-            (2, 4): axelrod.Match((axelrod.Defector(), axelrod.GoByMajority()), turns=turns),
-            (3, 3): axelrod.Match((axelrod.Grudger(), axelrod.Grudger()), turns=turns),
-            (3, 4): axelrod.Match((axelrod.Grudger(), axelrod.GoByMajority()), turns=turns),
-            (4, 4): axelrod.Match((axelrod.GoByMajority(), axelrod.GoByMajority()), turns=turns),
-        }
-        tournament._play_matches(matches)
-        for m in matches.values():
-            self.assertEqual(len(m), turns)
-            # Check that the name of the winner is in the list of names of
-            # players in the tournament (the way the matches are created these
-            # are not the same players).
-            self.assertIn(str(m.winner()), [str(p) for p in tournament.players] + ['False'])
+
+        player_classes = [axelrod.Cooperator, axelrod.TitForTat,
+                          axelrod.Defector, axelrod.Grudger]
+        matches = []
+        for i, player_cls in enumerate(player_classes):
+            for j, opponent_cls in enumerate(player_classes):
+                players = (player_cls(), opponent_cls())
+                match = axelrod.Match(players, turns=turns)
+                matches.append(((i, j), match))
+        matches_generator = (element for element in matches)
+
+        self.assertEqual((len(list(matches_generator))), len(matches))
+
+        interactions = tournament._play_matches(matches_generator)
```

Comment:
> I'm surprised that this generator works again -- shouldn't it be empty after the above assertion?

Yeah that is weird! 

I've just seen what happens when I add:

```
411         list(matches_generator)
412         self.assertEqual((len(list(matches_generator))), len(matches))
```

I then get the expected failure:

```
    lambda data: test(*arguments, **example_kwargs)
  File "/Users/vince/src/Axelrod/axelrod/tests/unit/test_tournament.py", line 412, in test_play_matches
    self.assertEqual((len(list(matches_generator))), len(matches))
AssertionError: 0 != 16
```

Does assert check "what would happen"? I don't know here. I've searched around and can't find anything immediately about this.

I've played around with it adding:

```
 410
 411         #self.assertEqual((len(list(matches_generator))), len(matches))
 412
 413         L = (n for n in range(5))
 414         self.assertEqual(len(list(L)), 5)
 415         M = [k for k in L]
 416         self.assertEqual(list(M), [0,1,2,3,4])
```

Note that I comment out 411 (this doesn't actually fail the tests) but adding 413 and 416 does, it gives:

```
F.........
======================================================================
FAIL: test_play_matches (axelrod.tests.unit.test_tournament.TestTournament)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vince/src/Axelrod/axelrod/tests/unit/test_tournament.py", line 392, in test_play_matches
    @example(turns=3)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/hypothesis/core.py", line 266, in wrapped_test
    lambda data: test(*arguments, **example_kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/hypothesis/executors.py", line 57, in default_new_style_executor
    return function(data)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/hypothesis/core.py", line 266, in <lambda>
    lambda data: test(*arguments, **example_kwargs)
  File "/Users/vince/src/Axelrod/axelrod/tests/unit/test_tournament.py", line 416, in test_play_matches
    self.assertEqual(list(M), [0,1,2,3,4])
AssertionError: Lists differ: [] != [0, 1, 2, 3, 4]

Second list contains 5 additional elements.
First extra element 0:
0

- []
+ [0, 1, 2, 3, 4]
```

Note that adding:

```
  417
  418         interactions = tournament._play_matches(matches_generator)
  419         self.assertEqual(len(interactions), len(matches))
```

(and removing those extra random tests) passes... I'm confused by this...

For the purpose of readability of the test, we could always just remove `self.assertEqual((len(list(matches_generator))), len(matches))`? It's not a terribly useful test, it was more for me to check the generator from list syntax had worked as I expected.

> 
> Also is there a reason to use a generator here? You've already got an iterable with matches. I could certainly see making matches in to a generator function instead of a list.

I only create a generator to make sure that the `_play_matches(matches_generator)` does indeed work on a generator and doesn't require a list.

Building the list and turning that in to the iterator seemed the simplest way to go...